dkpro-core-io-bincas-asl/src/main/java/de/tudarmstadt/ukp/dkpro/core/io/bincas/BinaryCasReader.java

```diff
@@ -76,54 +97,71 @@ public void getNext(CAS aCAS)
-                if (header[0] == (byte) 0xAC && header[1] == (byte) 0xED) {
-                    CASMgrSerializer casMgr = readCasManager();
-                    ((CASImpl)aCAS).setupCasFromCasMgrSerializer(casMgr);
-                }
+            	xts = readTypeSystem(header, aCAS);
             }
             
-            // BEGIN -- Legacy DKPro file format support
-            // Check if this is original UIMA CAS format or DKPro Core format
-            if (Arrays.equals(header, DKPRO_HEADER)) {
-                // If it is DKPro Core format, read the type system
-                getLogger().debug("Found DKPro-Core-style embedded type system");
-                ObjectInputStream ois = new ObjectInputStream(bis);
-                CASMgrSerializer casMgr = (CASMgrSerializer) ois.readObject();
-                if (ts == null) {
-                    ts = casMgr.getTypeSystem();
-                    ts.commit();
-                }
-            }
-            else {
-                // No embedded TS, reset 
-                bis.reset();
-            }
-            // END -- Legacy DKPro file format support
+            xts = readDKProHeader(bis, header, xts);
             
-            SerialFormat format;
-            if (ts != null) {
-                format = CasIOUtils.load(bis, aCAS, ts);
-            }
-            else {
-                format = CasIOUtils.load(bis, aCAS);
-            }
-            getLogger().debug("Found format " + format);
-        }            
-        catch (ClassNotFoundException e) {
-            throw new IOException(e);
+            if (xts != null) {
+				tsd2 = TypeSystemUtil.typeSystem2TypeSystemDescription(xts);
+			} else {
+				CasIOUtils.load(bis, null, aCAS, CasLoadMode.REINIT);
+				tsd2 = TypeSystemUtil.typeSystem2TypeSystemDescription(aCAS.getTypeSystem());
+			}
         }
         
+	        try {
+				// Merge the two typesystems
+				TypeSystemDescription mergedTypeSystem = CasCreationUtils.mergeTypeSystems(Arrays
+						.asList(TypeSystemUtil.typeSystem2TypeSystemDescription(typeSystem), tsd2));
+	
+				// Create a new CAS based on the merged type system
+				JCas mergedTypeSystemCas = CasCreationUtils.createCas(mergedTypeSystem,
+						(TypePriorities) null, (FsIndexDescription[]) null).getJCas();
+	
+				// Create a holder for the CAS metadata
+				CASMgrSerializer casMgrSerializer = Serialization
+						.serializeCASMgr((mergedTypeSystemCas).getCasImpl());
+	
+				// Reinitialize CAS with new type system
+				((CASImpl) aCAS).setupCasFromCasMgrSerializer(casMgrSerializer);
+	
+			} catch (CASException | ResourceInitializationException e) {
+				throw new CollectionException(e);
+			}
+		}
+		
+		// Read file again, this time into a CAS which has been prepared with the merged TS
+		try (InputStream is = CompressionUtils.getInputStream(res.getLocation(),
+				res.getInputStream())) {
+			BufferedInputStream bis = new BufferedInputStream(is);
+			bis.mark(32);
```

Comment:
the mark() is called here, but the reset() call is hidden in readDKProHeader. mark and reset should both be called in the same method to make the code more understandable.